ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto

```diff
@@ -57,11 +58,37 @@ message Commands {
   // Required
   repeated Command commands = 8;
 
-  // The length of the time window during which all commands with the same ``act_as`` parties and
-  // the same command ID will be deduplicated.
-  // Duplicate commands submitted before the end of this window return an ALREADY_EXISTS error.
+  // Specifies the deduplication period for the change ID.
+  // Deduplication provides the following deduplication guarantee:
+  // If the completion for this submission is a definite answer, say at the completion offset `off`,
+  // then there is no accepting completion for the same change ID between the start of the deduplication period and `off`.
+  //
+  // The participant or Daml ledger may reject the submission with OUT_OF_RANGE
+  // if the deduplication period starts in the future (e.g., by specifying an offset that is higher than the completion end).
+  // It may reject the submission with NOT_FOUND
+  // if the deduplication period starts too early (e.g., by specifying an offset before the completion begin).
+  //
+  // If omitted, the participant chooses the start of the deduplication period at its discretion.
   // Optional
-  google.protobuf.Duration deduplication_time = 9;
+  oneof deduplication {
+    // Specifies the start of the deduplication period by a completion stream offset.
+    //
+    // Must be a valid LedgerString (as described in ``value.proto``).
+    string deduplication_offset = 14;
+
+    // Specifies a point in time at which the deduplication shall start.
+    // The participant or Daml ledger converts this timestamp into a completion offset using its local clock.
+    // The clock for this purpose is typically not fully synchronized with the application’s clock or record time,
+    // and message and processing delays may further increase the difference in interpretation.
+    google.protobuf.Timestamp deduplication_start = 15;
+
+    // Specifies the length of the deduplication period.
+    // The participant or Daml ledger converts the duration into a timestamp `deduplication_start` since
+    // by subtracting the duration from its local clock at some point during processing the submission.
+    //
+    // Must be non-negative.
+    google.protobuf.Duration deduplication_time = 9;
```

Comment:
The semantics of this field changes from extending into the future to looking into the past.

Reply:
I'd suggest to allow the participant to distinguish old and new semantics. This way it can provide proper log and error messages that inform the user about deprecation and defaulting. Moreover, users can choose the point at which they switch.

Reply:
I agree with the advantages of a separate field. The downside is that all applications must explicitly be migrated to the new API, which is different from what we had discussed in the architecture meeting. The obvious compromise is accumulating some tech debt, namely having both fields, deprecating the old one at some point and removing it only after the deprecation cycle. Changed.